Marlin/src/feature/bedlevel/ubl/ubl_G29.cpp

```diff
@@ -1298,6 +1302,55 @@ mesh_index_pair unified_bed_leveling::find_closest_mesh_point_of_type(const Mesh
   return closest;
 }
 
+#if ENABLED(UBL_HILBERT_CURVE)
+   void unified_bed_leveling::hilbert(mesh_index_pair &pt, float x, float y, float xi, float xj, float yi, float yj, int n) {
+      /* Hilbert space filling curve implementation
```

Comment:
Why do you use float here ?

Reply:
I tried int first. Didn’t work. The Hilbert implementation relies on floating point numbers. While it may be possible to convert it to fixed point numbers I didn’t really feel like doing it.

Reply:
I believe the only fractional number it uses is 0.5, so one could use a single bit for the decimals. But again, this would require some work.

Reply:
No worry, this is a very good work already. The concerns I have is that it's a recursive function with the Hilbert level, so it'll overflow the stack if the subdivision is too high, and many (if not most) platform don't support floating point natively (and Hilbert's curves are typically implemented with fixed point usually, since, as you said, it's a clever division by a 1/4).

I've once implemented a space partitioning system based on Hilbert, that was using this trick to avoid using floating point value and fixed point value:
```
        /** This method was used to test the method above.
            It is twice faster than AJ Yzelman's thesis version, because of shifts instead of divisions */
        inline int computeHilbertsCurveValueFast(int x, int y, int logLevel)
        {
            static int HilbertRotationTable[4]  = { 3, 0, 0, 1};
            static int HilbertSenseTable[4]     = {-1, 1, 1, -1};
            static int HilbertQuadTable[4][2][2]= { {{0,1},{3,2}},
                                                    {{1,2},{0,3}},
                                                    {{2,3},{1,0}},
                                                    {{3,0},{2,1}} };


                int         rotation = 0;
                int         sense    = 1;
                int         quad     = 0;

                int         num      = 0;
                int         k        = 0;
                int         xbit     = 0;
                int         ybit     = 0;


                for(k = logLevel - 1; k > 0; --k)
            {
                        xbit = x>>k;
                        ybit = y>>k;
                        x -= xbit<<k;
                        y -= ybit<<k;
                        quad = HilbertQuadTable[rotation][xbit][ybit];
                        num += ((sense == -1) ? (3-quad) : quad) << (k*2);
                        rotation += HilbertRotationTable[quad];
                        if (rotation >= 4) rotation -= 4;
                        sense *= HilbertSenseTable[quad];
                }
                return num;
        }
```

Reply:
Wow. This implementation is even more difficult to understand than the one I used. I wouldn’t mind it, except that it goes the wrong way. It generates an 1D coordinate from a 2D coordinate. I need the opposite.

I understand your concern about blowing the stack. I wrote this for a Due based board, it may not be appropriate for 8-but micros. I can probably convert to fixed point numbers somewhat easily. That would partially mitigate the issue. I’ll see if I can do that.

Reply:
(Don't try to understand the output here, as I was generating an infinite length index using 2 bits per Hilbert Level). The trick I was try to show here is to shift left, perform the Hilbert curve 90° rotation and shift right just after to find the next coordinate, so that the coordinate can still be a integer.

Reply:
Ah. Okay. I get it. I’ve used fixed point math a lot in my life. I may write a to_fix and from_fix constexpr function for legibility, though.